test/jdk/jdk/incubator/vector/benchmark/src/main/java/benchmark/bigdata/VectorDistance.java

```diff
@@ -0,0 +1,541 @@
+    static float[] inputVectorFloat = parseBase64ToVector(x2);
+    static double[] queryVectorDouble;
+    static double[] inputVectorDouble;
+
+    static float normQueryVectorFloat;
+    static double normQueryVectorDouble;
+
+    public static float[] parseArray(byte[] input) {
+        if (input == null) {
+            return null;
+        }
+        float[] floatArr = new float[input.length / 4];
+        for (int i = 0; i < floatArr.length; i++) {
+            int l;
+            l = input[i << 2];
+            l &= 0xff;
+            l |= ((long) input[(i << 2) + 1] << 8);
+            l &= 0xffff;
+            l |= ((long) input[(i << 2) + 2] << 16);
+            l &= 0xffffff;
+            l |= ((long) input[(i << 2) + 3] << 24);
+            floatArr[i] = Float.intBitsToFloat(l);
+        }
+        return floatArr;
+    }
+
+    public static float[] parseBase64ToVector(String vectorBase64) {
+        return parseArray(Base64.getDecoder().decode(vectorBase64));
+    }
+
+    @Setup
+    public static void init() {
+        queryVectorDouble = new double[queryVectorFloat.length];
+        inputVectorDouble = new double[inputVectorFloat.length];
+        for (int i = 0; i < queryVectorFloat.length; i++) {
+            queryVectorDouble[i] = (double)(queryVectorFloat[i]);
+        }
+        for (int i = 0; i < inputVectorFloat.length; i++) {
+            inputVectorDouble[i] = (double)(inputVectorFloat[i]);
+        }
+        float xSquare = 0;
+        for (int i = 0; i < queryVectorFloat.length; i++) {
+            xSquare += (float)(queryVectorFloat[i] * queryVectorFloat[i]);
+        }
+        normQueryVectorFloat = xSquare;
+        normQueryVectorDouble = (double)xSquare;
+    }
+
+    @Benchmark
+    public static float cosinesimilOptimizedScalarFloat() {
+        float dotProduct = 0.0f;
+        float normInputVector = 0.0f;
+        for (int i = 0; i < queryVectorFloat.length; i++) {
+            dotProduct += queryVectorFloat[i] * inputVectorFloat[i];
+            normInputVector += inputVectorFloat[i] * inputVectorFloat[i];
+        }
+        float normalizedProduct = normQueryVectorFloat * normInputVector;
+        if (normalizedProduct == 0) {
+            return Float.MIN_VALUE;
+        }
+        return (float) (dotProduct / (Math.sqrt(normalizedProduct)));
+    }
+
+    @Benchmark
+    public static float cosinesimilOptimizedVectorFloat128() {
+        VectorSpecies<Float> SPECIES = SPECIES_FLOAT_128;
+        FloatVector vecX, vecY, vecSum, xSquareV, ySquareV;
+        vecSum = FloatVector.zero(SPECIES);
+        xSquareV = FloatVector.zero(SPECIES);
+        ySquareV = FloatVector.zero(SPECIES);;
+        int i= 0;
+        for (i = 0; i + (SPECIES.length()) <= queryVectorFloat.length; i += SPECIES.length()) {
+            vecX = FloatVector.fromArray(SPECIES, queryVectorFloat, i);
+            vecY = FloatVector.fromArray(SPECIES, inputVectorFloat, i);
```

Comment:
Thanks for your JMH benchmarks! Could you please use `"SPECIES_FLOAT_128"` inside the API instead of defining a new variable here? I'm wondering whether the API can be vectorized by using `"SPECIES"` here, since the API needs the SPECIES argument to be a constant. I'm not sure about this case, but I really met the same issue before when I ran the jtreg tests. To make sure everything works as expect,  I'd like using the `static final field "SPECIES_FLOAT_128"` all through the API usage.

Reply:
done